### ISSUE TITLE
Support multiple (partial) images per canvas

### ISSUE THREAD

**Issue by azaroth42** [ProjectMirador/mirador#632]
Support overlay of multiple images on a canvas, with z order from the first (lowest) to last (highest).

Understood that this is dependent on OSD collections mode, but want to make sure the explicit functionality is available.

**Comment by beaudet** [comment]
For ConservationSpace, we have an overlay requirement as well, and although I was initially thinking it would best be provided by overlaying images on a single canvas, I now think that overlaying slots might work better in our case due to the complexity of the authoring tools that would be required to support separate pan/zoom of images on a canvas in order to get them all lined up.  I'll list the issue number here once I create an issue for the conservation space overlay feature.

**Comment by azaroth42** [comment]
@beaudet - Once the authoring is done, the images would be aligned on a single canvas for rendering? In other words, to separate the UI from the Presentation API representation, there would be two modes:

Authoring: New issue to come
Rendering: This issue

Yes?

**Comment by beaudet** [comment]
I think it will become more clear when I post the use case with some mock-ups, but the authoring actually takes place using Mirador.  Two images of the same subject that may or may not be identical resolutions and origins need to be aligned with each other manually.  Seems like quite an enhancement to Mirador to support adjustment of the zoom level and pan settings of individual images on a canvas.  Whereas if the same canvas was loaded into two slots, and the slots combined via tabs in the same slot window, the slots could be overlaid on each other, transparency controlled, pan / zoom  adjusted until the overlay was achieved.  Then, the slots could have their pan/zoom controls locked to each other which would permit inspection of the two slots simultaneously with an overlay.  The nice thing about this ad-hoc image overlay is it's performed almost exclusively with Mirador features.  The downside is that there's probably no way to represent this in the presentation API, so it might have to be reverse engineered into its own canvas definition to reconstruct the view - or it becomes a Mirador specific configuration that is persisted which I would like to try to avoid if possible when saving such views because it will be difficult to sustain / reconstitute those in the future when newer versions of Mirador are released.

**Comment by beaudet** [comment]
Rob, I had originally imagined that overlay images would be part of the same canvas as well (makes more sense since the protocol allows for multiple images per canvas, etc.) but I also wanted to avoid the authors having to define images as a registered set outside of Mirador using the ConservationSpace application or another external system.  Defining them in advance also makes ad-hoc comparisons less... well, ad-hoc.  

I mulled this problem over again today and I think there's a reasonable user experience that could be constructed within Mirador to allow canvases to be overlaid on each other within a single slot without having to actually represent the images on the same canvas.  This preserves the ad-hoc nature of the overlay feature that also requires those images to be independently panned and zoomed.    In cases where metadata is available defining image sets that are actually registered, a manifest could certainly be created with multiple images on the same canvas, although placing more than two images on the same canvas could get a bit confusing without adequate transparency controls in place, but I see that as a future enhancement that could build upon the initial work of overlaying two canvases.

Extending what might be a cleaner data model approach (multiple overlaid images on a single canvas) into a UI that allows independent pan / zoom for each image sounds more like a full manifest editor which would be awesome, but is not something that I think we have time for within the ConservationSpace project.  Not having independent pan/zoom for images on a single canvas means the ad-hoc overlay user story is probably not achievable in the short term.

However, there might also be a downside to this approach since a canvas might represent a number of images "stitched" together, e.g. an altarpiece reconstruction, so you would potentially face a situation like this: you have 24 images on the same canvas, 12 visible light images that are assembled like puzzle pieces to reconstruct an altarpiece and 12 x-ray images similarly assembled and you want to overlay the first set of 12 on the second set of 12.  It's not clear to me how that would be modeled on the same canvas.  Putting 12 on one canvas and 12 on another might be easier.

Anyway, it seems like there are two paths here and they could actually both be followed, but I think the important question to answer is whether people think that the concept of overlaying two canvases on each other with transparency control makes sense or not.  In my mind, this could consist of a UI capability to "pin" one of the canvases of a manifest on top of the canvas being viewed - essentially creating a single transparency layer of the pinned canvas.  The pan/zoom control would default to unlocked when a "pinned" canvas had been selected, but could be locked as well to enable the two canvases to be scrolled and zoomed as one.  I imagine the data for this state could be represented as a few pieces of metadata for the canvas marked as "pinned".

The other path amounts to modeling registered images sets using the existing protocols and I  imagine that one could create three different canvases for a two-image overlay using perhaps 30%, 50%, and 70% transparency blends and users could just select the blend that is most useful to them.  Eventually, UI controls could be created that allow these structures to be created directly in Mirador with the location, zoom level, transparency, rotation, brightness, etc. set for each image on a canvas.

Does this make sense or am I way off the mark with the vision for the Mirador tool?

**Comment by beaudet** [comment]
Drew, Rashmi and I came to agreement on an initial design for displaying a list of layers on a canvas as well as providing for an ability to add new layers using any any image resource in the manifest.  For now, the image resource list will be optionally configurable (off by default) since the initial, non-filtered, non-paginated list of images resources would be a performance concern for manifests with many images.  @rsinghal can you please review the following specifications and assign to @stanislavkostadinov?  I am including links to two documents @aeschylus, one an abbreviated simple (default) behavior, and one using both "alternate depictions" when in the presence of oa:annotation choices, and also the image resource list turned on so that new layers can be created from them.  @mikeapp, I don't think @stanislavkostadinov should build in support for the alternate depictions since we don't have a need for those yet in ConservationSpace, although I've built it into the design.

The simple design doc: https://drive.google.com/file/d/0B8dnsz8s0G6eajUxMU5iVmFMWmc/view?usp=sharing

The less simple design doc that @stanislavkostadinov will need to target in order to support ad-hoc overlays for ConservationSpace: https://drive.google.com/file/d/0B8dnsz8s0G6eT2lIVjNCNm9LcWs/view?usp=sharing

and the specifications:

Layers Dialog for Overlay Support
ConservationSpace has a requirement to allow images to be overlaid on top of one another and the opacity controlled so that, for example, an infrared image can be overlaid on top of a visible light image and then transformed so that its size can be adjusted to permit alignment of the two images with subsequent adjustment of the opacity of either image in order to provide visual cues for where, on the physical object, the infrared light was transmitted.

General Feature Design
•	A re-orderable list of images appearing on the selected canvas shall be rendered in a scrolling “Layers” tab in Z index that includes a thumbnail of the image resource and its label.
•	In the Mirador 2.1 branch, an empty, hidden “layers” tab has been committed at https://github.com/IIIF/mirador/blob/release2.1/js/src/widgets/layersTab.js and you can enable it here: https://github.com/IIIF/mirador/blob/release2.1/js/src/workspaces/window.js#L46
•	Images appearing in the “Layers” tab can be reordered by dragging and dropping them into position.  Such changes are considered to be immediately committed, i.e. there is no cancel – the user would just have to revert to the prior order if they mess up.
•	Images appearing in the “Layers” tab may be deleted from the canvas if they are "unprotected images"
•	As shown in the design document, a dropdown arrow will appear next to each layer in the list.  Clicking on this icon will cause the system to enter “layer editing” mode.
     o	In layer editing mode, the image layer is temporarily brought to the foreground if it's not already and adorned with the same set of handles and rotation tools used for rectangles (but without a delete icon).  By moving these handles, image layers can be dilated, translated, and rotated, but the aspect ratio should not be able to be changed.  In addition, a set of editing controls for the layer appear in a dialog box below the line item. In the dialog, a set of sliders appears for controlling the layer’s opacity, contrast, and brightness, each with corresponding floating point text boxes next to the slider - we might also want to consider allowing the label to be edited, but it's not a requirement for ConservationSpace. Save and cancel buttons appear in the editing dialog that result in either discarding the changes or committing them in memory, including any transformations made to the image layer using the handles.  Upon saving or canceling, the mode reverts to viewing mode.
     o	When a layer is deleted, a confirmation dialog confirms the deletion before it takes effect.
•	When image layers are added, deleted, or edited, the Manifest within Mirador shall be updated immediately with the new configuration and, after the merge of the IIIF Manifests Layout component, OSD will be called to re-render this modified manifest.  
•	Since a new Manifesto-like API will not be part of the Mirador 2.1 release, the advised method for editing the manifest after image layer manipulations is to use the existing internal manifest representations of Mirador rather than regenerating the manifest externally and causing Mirador to reload it from scratch every time a layer is edited, added, or deleted.
•	Below the list of layers is a list of all image resources defined for the canvas.  The list of image resources is formatted similarly to the list of layers, but without any action controls.  Dragging one of these image resource into the layers list creates a new layer in the desired position within the list of layers.  An image resource may be duplicated multiple times in this manner.  This feature should be turned OFF by default but may be configured to appear if desired.

Annotation Modelling Differences between External Systems and IIIF Manifests
For some applications that embed Mirador, it’s important to consider that annotations might be mapped directly to image resources with no concept of a “canvas”.  While translation glue between IIIF and the external system must be written to fetch and store annotations like this, it's likely to be a typical scenario.  

ConservationSpace is an external system of this type. It has no model of a canvas, generates Manifests dynamically from an image widget configuration, and stores annotations with the image for which a canvas was initially created.  Due to these modelling differences, ConservationSpace does not want to permit deletion of these initial images and would like to bless them as “protected” images.  

ConservationSpace will therefore add a capability in Mirador for an external API to designate one or more images per canvas as “protected”.  Protected images will not be able to be deleted from the layers list nor transformed, although brightness, contrast, and opacity should be able to be modified as should the ordering of the protected image in the list of layers.  Since this status cannot be modeled with the current version of IIIF Manifests, the blessing of “protected images” must be called separately by an external application after a canvas, or perhaps a manifest, is loaded.

**Comment by azaroth42** [comment]
And to make certain I understand the model implications ... this will trigger from the presence of an oa:Choice of images?

**Comment by aeschylus** [comment]
Choice and "detail" (images that are children of the canvas). 

See: https://github.com/sul-dlss/iiifManifestLayouts/issues/40#issuecomment-159741392
https://github.com/sul-dlss/iiifManifestLayouts/issues/98#issuecomment-161451617

**Comment by azaroth42** [comment]
> the image resource list will be optionally configurable (off by default)

You're defaulting detail images to off?

**Comment by beaudet** [comment]
There is a second list below the canvas layers list that contains all image resources from the manifest, regardless of the canvas on which they appear.  Any item from that list can be dragged into the layers list to create a new layer on the canvas being viewed.  It is this list of image resources from the Manifest that is optional. It's optional for now because it would require a filter + pagination to make it useful for manifests with a great number of image resources.

**Comment by azaroth42** [comment]
For authoring then, sorry for the confusion!

> Authoring: New issue to come
> Rendering: This issue

This would help!

**Comment by beaudet** [comment]
Yeah, I was wondering if I was hijacking this issue.  I'll create a new one for the layers tab.

**Comment by aeschylus** [comment]
@azaroth42 We're not sure what the defaults should be. It seems rational that the most basic version of the image resources/choice sidepanel is always on, and that all base/detail images are drawn on the canvas by default, with alternates hidden. Clicking the "eyeball" icon on those resources would display them in place of the "base" image of the choice block.

**Comment by azaroth42** [comment]
Yep 👍 :)

**Comment by imoutsatsos** [comment]
I want to add my 2cents here as well, that from scientific imaging perspective the ability to overlay multi-spectral images is essential. I'm working with images from cell microscopy that are typically imaged at different wavelengths to detect fluorescently labeled cellular markers. At the end these separate gray scale images need to be pseudo-colored and overlayed to create a single color micrograph.

As an example, the following micrograph displays a pseudo color image from 3 different fluorescent stains (blue: nuclei, red: cytoplasm, green:RNA)
![image](https://cloud.githubusercontent.com/assets/3408226/24419933/fe44e4d6-13be-11e7-951b-b59f120a38a9.png)

Thanks for your open source contributions!!
Regards
Ioannis

**Comment by aeschylus** [comment]
@imoutsatsos This is great. Do you have a manifest for these objects modelling this with "oa:choice"? If so, could I have a link to use it as a test object? We're planning to merge this functionality this week.

**Comment by imoutsatsos** [comment]
@aeschylus I'm afraid I'm new to creating manifests. I would be happy to create one if you can provide me with some relevant instructions. I have upload the relevant images here: https://github.com/imoutsatsos/iiif-image-resources

**Comment by mejackreed** [comment]
Closing this as its supported in Mirador 3 already.
